Deprecate window global

NullVoxPopuli opened the pull request.

rwjblue commented on packages/ember/index.js:
```diff
@@ -171,7 +171,7 @@ import {
 
 Ember.isNamespace = true;
-Ember.toString = function () {
+Ember.toString = function() {
```
you'll want to roll back these unrelated prettier changes (also seems like maybe your editor was applying a different prettier version than our linting uses)

NullVoxPopuli replied:
for sure, idk what happened here (very WIP PR) :p

rwjblue commented on packages/ember/index.js:
```diff
@@ -457,7 +457,7 @@ Ember.isEqual = isEqual;
   assert(
     `Injected properties must be created through helpers, see '${Object.keys(inject)
-      .map((k) => `'inject.${k}'`)
+      .map(k => `'inject.${k}'`)
```
🥒

rwjblue commented on packages/ember/index.js:
```diff
@@ -647,11 +647,11 @@ Ember.HTMLBars = {
 
 if (ENV.EXTEND_PROTOTYPES.String) {
-  String.prototype.htmlSafe = function () {
+  String.prototype.htmlSafe = function() {
```
🍉

rwjblue commented on packages/ember/index.js:
```diff
@@ -647,11 +647,11 @@ Ember.HTMLBars = {
   };
 }
-const deprecateImportFromString = function (
+const deprecateImportFromString = function(
```
🍅

rwjblue commented on packages/ember/index.js:
```diff
@@ -791,6 +791,36 @@ if (IS_NODE) {
   context.exports.Ember = context.exports.Em = Ember;
 }
 
+if (window) {
+  Object.defineProperty(window, 'Ember', {
+    configurable: true,
+    writable: true,
+    get() {
+      deprecate(
+        "Using window.Ember has been deprecated. Use `import Ember  from 'ember';` instead ",
+        false,
+        {
+          id: 'ember-source.window-global',
+          until: '4.0.0',
+          url: 'tbd',
+          for: 'ember-source',
+          since: { enabled: '3.26.0-beta.1' },
+        }
+      );
+
+      Object.defineProperty(window, 'Ember', {
+        configurable: true,
+        writable: false,
+        get() {
+          return Ember;
+        },
+      });
+
+      return Ember;
+    },
+  });
+}
```
I think this needs to be just above where `context.exports.Ember = context.exports.Em = Ember;` is.

`context.exports` _is_ `window` IIRC

NullVoxPopuli replied:
oh! this is good to know. thanks!

rwjblue reviewed (changes requested):
I think this will also need a couple of tests

pzuraq commented:
So had a quick convo with @rwjblue about this, and we realized we're going to need to make some additional changes. Currently, `ember-cli-babel` still transpiles `import * from '@ember/*'` to use the Ember global, and it does _not_ add `import Ember from 'ember';` at the top. In addition, we don't actually _define_ an `ember` module - that is also transpiled via `ember-cli-babel`.

So, in this package we need to add the Ember package so you can actually import it. In the babel plugin that `ember-cli-babel` adds, we need to update it to use `import Ember from 'ember';` to the top. It should also only do that IF the Ember version is high enough, so probably 3.26.0-beta.1 assuming we get this PR in before then.

NullVoxPopuli commented:
Next Steps?:
- [ ] update ember-cli-babel
- [ ] add deprecation guide
- [ ] merge this PR?

rwjblue reviewed (changes requested):
In order to land we need to:

* Update https://github.com/ember-cli/babel-plugin-ember-modules-api-polyfill 
  * Add a new configuration option (`useEmberModule` maybe?) to opt in to emitting the `import Ember from 'ember';` bit or doing the (currently default) "just use `window.Ember`" thing
* Update ember-cli-babel
  * to use that new version of babel-plugin-ember-modules-api-polyfill
  * to pass `true` or `false` to that new option based on the running version of `ember-source`
* Update this PR to actually emit a `ember` module (current Ember apps actually have no `ember` module at runtime, we have to fix that)

btecu commented:
Aren't there some important parts that have not been exposed via modules yet? How would that work with this?
For example `Ember.onError`, is used for error trackers such as Bugsnag and Sentry.

pzuraq commented:
@btecu as noted in the RFC, you will still be able to `import Ember from 'ember';`. This is just about accessing the object without importing it.

NullVoxPopuli commented:
babel plugin: https://github.com/ember-cli/babel-plugin-ember-modules-api-polyfill/pull/175

pzuraq commented:
@NullVoxPopuli can you rebase this PR? Now that #19390 has landed, everything should be much easier to deprecate since the global is now setup lazily (you can do it in `@ember/-internals/bootstrap`). Also we need to add a test that verifies the deprecation runs when we use the global value

NullVoxPopuli committed:
[RFC#706]: Deprecate window.Ember and window.Em, which is an alias of window.Ember

RFC: emberjs/rfcs#706

tl;dr: Recommend `import Ember from 'ember';` instead of using the window.Ember
global

rwjblue commented:
https://github.com/ember-cli/babel-plugin-ember-modules-api-polyfill/releases/tag/v3.3.0 includes new `useEmberModule` option. Next up: update ember-cli-babel.

simonihmig commented:
This has been superseded by https://github.com/emberjs/ember.js/pull/19457, so can be closed, right?
